File: samples/python/src/GraphMapNodes.py
from __future__ import annotations
from typing import Any, TypeVar, Generic

T = TypeVar("T")


class GraphMapNodes:

    @staticmethod
    def run_main():
        GraphMapNodes().start()

    def start(self):

        # graph with hashmap node
        gmap: dict[
            GraphMapNodes.GraphMapNode[str], list[GraphMapNodes.GraphMapNode[str]]
        ] = {}
        gedges: dict[str, int] = {}
        Extractor.gedges = gedges

        # nodes
        node1 = GraphMapNodes.GraphMapNode[str]("1")
        node2 = GraphMapNodes.GraphMapNode[str]("2")
        node3 = GraphMapNodes.GraphMapNode[str]("3")
        node4 = GraphMapNodes.GraphMapNode[str]("4")
        gmap[node1] = [node2, node3]
        gmap[node2] = [node3, node4]
        gmap[node3] = []
        gmap[node4] = []

        # edges
        gedges[(node1, node2)] = 100
        gedges[(node1, node3)] = 200
        gedges[(node2, node3)] = 300
        gedges[(node2, node4)] = 400

        print("done")

    class GraphMapNode(Generic[T]):
        def get_value(self) -> T:
            return self.value

        def __init__(self, value: T):
            self.value = value

        def __str__(self):
            return str(self.value)


# custom extractor
class Extractor:
    # register the custom attributes
    @staticmethod
    def register_attrs() -> list[list[str, list[str]]]:
        return [
            ["dict", ["dictCustomNodes", "dictCustomValue"]],
            ["GraphMapNode", ["nodeCustomNodes", "nodeCustomEdges", "nodeCustomValue"]],
        ]

    @staticmethod
    def extract_dictCustomNodes(
        obj: dict[
            GraphMapNodes.GraphMapNode[str], list[GraphMapNodes.GraphMapNode[str]]
        ],
        root: dict[
            GraphMapNodes.GraphMapNode[str], list[GraphMapNodes.GraphMapNode[str]]
        ],
    ) -> list[GraphMapNodes.GraphMapNode[str]]:
        nodes: list[GraphMapNodes.GraphMapNode[str]] = []
        for key in root:
            nodes.append(key)
        return nodes

    @staticmethod
    def extract_dictCustomValue(
        obj: dict[
            GraphMapNodes.GraphMapNode[str],
            list[GraphMapNodes.GraphMapNode[str]],
        ],
        root: Any
    ) -> list[str]:
        sb = ""
        for key in obj.keys():
            if len(sb) > 0:
                sb += ","
            sb += key.get_value()
        return [sb]

    @staticmethod
    def extract_nodeCustomNodes(
        obj: GraphMapNodes.GraphMapNode[str],
        root: dict[
            GraphMapNodes.GraphMapNode[str],
            list[GraphMapNodes.GraphMapNode[str]],
        ],
    ) -> list[GraphMapNodes.GraphMapNode[str]]:
        return root.get(obj, [])

    @staticmethod
    def extract_nodeCustomEdges(
        obj: GraphMapNodes.GraphMapNode[str],
        root: dict[
            GraphMapNodes.GraphMapNode[str],
            list[GraphMapNodes.GraphMapNode[str]],
        ],
    ) -> list[str]:
        edges: list[int] | None = []
        for child in root.get(obj, []):
            edgeKey = (obj, child)
            edgeValue: int | None = Extractor.gedges.get(edgeKey)
            if edgeValue:
                edges.append(edgeValue)
        return edges

    @staticmethod
    def extract_nodeCustomValue(
        obj: GraphMapNodes.GraphMapNode[str],
        root: Any
    ) -> list[str]:
        return [str(obj.get_value())]


GraphMapNodes.run_main()

File: samples/csharp/src/GraphMapNodes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

// UI settings
// Expression: gmap
//   Type: System.Collections.Generic.Dictionary<GraphMapNodes.GraphMapNode<string>, System.Collections.Generic.List<GraphMapNodes.GraphMapNode<string>>>
//     Layout: Graph
//     Nodes: @dictCustomNodes
//     Properties: @dictCustomValue (optional)
//     Save: Object Attributes
//   Type: GraphMapNodes.GraphMapNode<string>
//     Layout: Graph
//     Nodes: @nodeCustomNodes
//     Edges: @nodeCustomEdges
//     Properties: @nodeCustomValue (optional)
//     Save: Object Type Attributes
internal class GraphMapNodes
{

    public static void RunMain(string[] args)
    {
        new GraphMapNodes().Start();
    }

    public void Start()
    {
        // graph with hashmap node
        Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> gmap
            = new Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>>();
        Dictionary<string, int> gedges = new Dictionary<string, int>();
        Extractor.gedges = gedges;

        // nodes
        GraphMapNode<string> node1 = new GraphMapNode<string>("1");
        GraphMapNode<string> node2 = new GraphMapNode<string>("2");
        GraphMapNode<string> node3 = new GraphMapNode<string>("3");
        GraphMapNode<string> node4 = new GraphMapNode<string>("4");

        gmap.Add(node1, new List<GraphMapNode<string>>([node2, node3]));
        gmap.Add(node2, new List<GraphMapNode<string>>([node3, node4]));
        gmap.Add(node3, []);
        gmap.Add(node4, []);

        // edges
        gedges.Add(node1.Value + "," + node2.Value, 100);
        gedges.Add(node1.Value + "," + node3.Value, 200);
        gedges.Add(node2.Value + "," + node3.Value, 300);
        gedges.Add(node2.Value + "," + node4.Value, 400);

        Console.WriteLine("done");
    }

    class GraphMapNode<T>
    {
        public T Value { get; private set; }
        public GraphMapNode(T value)
        {
            this.Value = value;
        }
    }

    // custom extractor
    static class Extractor
    {
        internal static Dictionary<string, int> gedges;

        // register the custom attributes to extract
        // you can select these from the ui 
        // instead of modifying your objects
        public static object[] RegisterAttrs()
        {
            return [
                new object[]{"System.Collections.Generic.Dictionary<GraphMapNodes.GraphMapNode<string>, System.Collections.Generic.List<GraphMapNodes.GraphMapNode<string>>>",
                    new string[]{"dictCustomNodes", "dictCustomEdges", "dictCustomValue"}},
                new object[] {"GraphMapNodes.GraphMapNode<string>",
                    new string[]{"nodeCustomNodes", "nodeCustomEdges", "nodeCustomValue"}}
            ];
        }

        public static List<GraphMapNode<string>> Extract_dictCustomNodes(
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> obj,
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> root)
        {
            List<GraphMapNode<string>> nodes = new List<GraphMapNode<string>>();
            foreach (GraphMapNode<string> node in root.Keys)
                nodes.Add(node);
            return nodes;
        }

        public static List<int> Extract_dictCustomEdges(
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> obj,
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> root)
        {
            return null; // return null to hide edges
        }

        public static string[] Extract_dictCustomValue(
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> obj,
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> root)
        {

            StringBuilder sb = new StringBuilder();
            foreach (GraphMapNode<string> key in obj.Keys)
            {
                sb.Append(key.Value);
                sb.Append(",");
            }
            return [sb.ToString()];
        }

        public static List<GraphMapNode<string>> Extract_nodeCustomNodes(
            GraphMapNode<string> obj,
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> root)
        {
            List<GraphMapNode<string>> nodes = new List<GraphMapNode<string>>();
            if (root.ContainsKey(obj))
                nodes.AddRange(root[obj]);
            return nodes;
        }

        public static List<string> Extract_nodeCustomEdges(
            GraphMapNode<string> obj,
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> root)
        {
            List<string> edges = new List<string>();
            if (root.ContainsKey(obj))
            {
                foreach (GraphMapNode<string> child in root[obj])
                {
                    string edgeKey = obj.Value + "," + child.Value;
                    if (gedges.ContainsKey(edgeKey))
                    {
                        int edgeValue = gedges[edgeKey];
                        edges.Add(edgeValue.ToString());
                    }
                    else
                    {
                        edges.Add("");
                    }
                }
            }
            return edges;
        }

        public static string[] Extract_nodeCustomValue(
            GraphMapNode<string> obj,
            Dictionary<GraphMapNode<string>, List<GraphMapNode<string>>> root)
        {
            return [obj.Value];
        }
    }
}
